zng/walk.go

```diff
@@ -195,85 +125,31 @@ func walkUnion(typ *TypeUnion, body zcode.Bytes, rv RecordVisitor) error {
 	if err != nil {
 		return err
 	}
-	switch t := AliasedType(inner).(type) {
-	case *TypeRecord:
-		if !container {
-			return &RecordTypeError{Name: "<record element>", Type: t.String(), Err: ErrNotContainer}
-		}
-
-		if err := walkRecord(t, body, rv); err != nil {
-			return err
-		}
-	case *TypeArray:
-		if !container {
-			return &RecordTypeError{Name: "<array element>", Type: t.String(), Err: ErrNotContainer}
-		}
-		if err := rv(t, body); err != nil {
-			if err == SkipContainer {
-				return nil
-			}
-			return err
-		}
-
-		if err := walkVector(t, body, rv); err != nil {
-			return err
-		}
-	case *TypeSet:
-		if !container {
-			return &RecordTypeError{Name: "<set element>", Type: t.String(), Err: ErrNotContainer}
-		}
-		if err := rv(t, body); err != nil {
-			if err == SkipContainer {
-				return nil
-			}
-			return err
-		}
-
-		if err := walkSet(t, body, rv); err != nil {
-			return err
-		}
-	case *TypeUnion:
-		if !container {
-			return &RecordTypeError{Name: "<union value>", Type: t.String(), Err: ErrNotContainer}
-		}
-		if err := rv(t, body); err != nil {
-			if err == SkipContainer {
-				return nil
-			}
-			return err
-		}
-		if err := walkUnion(t, body, rv); err != nil {
-			return err
-		}
-	default:
-		if container {
-			return &RecordTypeError{Name: "<union value>", Type: t.String(), Err: ErrNotPrimitive}
-		}
-		if err := rv(t, body); err != nil && err != SkipContainer {
-			return err
-		}
+	if !it.Done() {
+		err := errors.New("union value container has more than two items")
+		return &RecordTypeError{Name: "<union>", Type: typ.String(), Err: err}
 	}
-	return nil
+	if err := checkKind("<union body>", inner, container); err != nil {
+		return err
+	}
+	return Walk(inner, body, visit)
 }
 
-func walkSet(typ *TypeSet, body zcode.Bytes, rv RecordVisitor) error {
+func walkSet(typ *TypeSet, body zcode.Bytes, visit Visitor) error {
 	if body == nil {
 		return nil
 	}
 	inner := AliasedType(InnerType(typ))
-	if IsContainerType(inner) {
```

Comment:
This is a behavior change... can you keep it for the subsequent work where you are lifting the primitive set type restriction, so that this commit remains refactor-only?

Reply:
Yes, good catch.  This was my intent and missed it here...